chore: [Booking Flow Refactor - 5] Move post booking things to separate service `BookingEventHandlerService` starting with HashedLink usage handling

hariombalhara opened the pull request:
## What does this PR do?

This PR refactors post-booking logic by extracting it from the main booking flow into a dedicated `BookingEventHandler` service. This is the first step in improving the post booking architecture by separating concerns and making the codebase more maintainable.

**Key Changes:**
- Extracts hashed link usage tracking from inline code into a new `BookingEventHandler` service
- Introduces type-safe event payloads for booking creation and rescheduling events
- Uses dependency injection everywhere
- Adds comprehensive test coverage for post-booking event handling to ensure no regression



## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

### Environment Setup
- Standard Cal.com development environment
- PostgreSQL database for hashed link functionality

### Test Scenarios
1. **Hashed Link Creation**: Create a booking with a hashed link and verify usage count increments
2. **Hashed Link Rescheduling**: Reschedule a booking with a hashed link and verify usage tracking
3. **Dry Run Handling**: Ensure dry runs don't increment usage counts
4. **Error Handling**: Verify that hashed link service errors don't break booking creation

### Expected Behavior
- All existing booking functionality should work unchanged
- Hashed link usage tracking should continue to work exactly as before
- Post-booking errors should be logged but not break the booking flow
- Tests in `packages/features/bookings/lib/handleNewBooking/test/post-booking-handling.test.ts` should pass

## Checklist

**⚠️ Critical Review Areas:**

- [ ] **Verify BookingEventHandler Integration**: Ensure the `BookingEventHandler` is properly instantiated and called in the booking flow (around line 2063 in `RegularBookingService.ts`)
- [ ] **Error Handling Strategy**: Confirm that changing from throwing `HttpError` to logging errors is the intended behavior for post-booking failures
- [ ] **Dependency Injection Configuration**: Validate that all DI modules are properly configured and dependencies are correctly wired
- [ ] **Service Location Changes**: Check that moving `HashedLinkService` doesn't break any existing imports or functionality
- [ ] **Test Coverage vs Implementation**: The tests simulate hashed link increments rather than fully mocking the service - verify this doesn't hide integration issues

**Architectural Changes:**
- [ ] New `BookingEventHandler` class properly handles both creation and rescheduling events
- [ ] Type definitions in `onBookingEvents/types.d.ts` are comprehensive and accurate
- [ ] Dependency injection setup follows Cal.com patterns and conventions
- [ ] Error boundaries prevent post-booking failures from affecting booking creation

**Backward Compatibility:**
- [ ] All existing booking flows continue to work unchanged
- [ ] Hashed link functionality maintains exact same behavior
- [ ] No breaking changes to public APIs or interfaces

---

**Link to Devin run**: https://app.devin.ai/sessions/4465fc388ef1433d80ca534e6c4ec2dd  
**PR Description Requested by**: @hariombalhara

**Note**: This refactoring provides a foundation for incrementally moving additional post-booking functionality (webhooks, notifications, analytics) into the event handler pattern in future PRs.

hariombalhara committed:
chore: Move post booking stuff to separate service

volnei commented on packages/features/bookings/lib/onBookingEvents/BookingEventHandlerService.ts:
```diff
@@ -0,0 +1,48 @@
+    await this.onBookingCreatedOrRescheduled(payload);
+  }
+
+  private async onBookingCreatedOrRescheduled(payload: BookingCreatedPayload | BookingRescheduledPayload) {
+    await Promise.allSettled([this.updatePrivateLinkUsage(payload)]);
+  }
+
+  private async updatePrivateLinkUsage(payload: BookingCreatedPayload | BookingRescheduledPayload) {
+    try {
+      if (payload.bookingFormData.hashedLink) {
+        await this.deps.hashedLinkService.validateAndIncrementUsage(payload.bookingFormData.hashedLink);
+      }
+    } catch (error) {
+      this.deps.log.error("Error while updating hashed link", safeStringify(error));
+    }
+  }
```
two observations here: 
1. I don't thing we need this `Promise.allSettled` unless we're returning a list of promises from `updatePrivateLinkUsage`

2. It's always better to avoid using `await` unless you immediately need this results (line 38)

volnei reviewed (dismissed):
I see we're introducing a new kind of DI which is good but thinking forward I prefer to not go this way right now to maturate a better structure (and place) for DI since lib folder I don't think is the right place.

hariombalhara commented on packages/features/bookings/lib/service/RegularBookingService.ts:
```diff
@@ -2045,6 +2047,42 @@ async function handler(
       }
     : undefined;
 
+  const bookingFlowConfig = {
+    isDryRun,
+  };
```
In one of my follow PRs I move, _bookingFlowConfig_ generation to validation service that validates all the data and returns a very readable format that has all the validated data.

hariombalhara commented on packages/features/bookings/lib/onBookingEvents/types.d.ts:
```diff
@@ -0,0 +1,39 @@
+  eventType: BookingEventType;
+  config: BookingFlowConfig;
+  bookingFormData: {
+    hashedLink: string | null;
```
More data will be added here later, as and when needed.

hariombalhara commented:
> I see we're introducing a new kind of DI which is good but thinking forward I prefer to not go this way right now to maturate a better structure (and place) for DI since lib folder I don't think is the right place.

Moved newly added _di_ modules to _features_ folder and moved existing ones as well that I touched.

hariombalhara commented on packages/features/bookings/lib/service/RegularBookingService.ts:
```diff
@@ -2227,10 +2229,47 @@ async function handler(
+    eventType,
+    config: bookingFlowConfig,
+    bookingFormData: {
+      // FIXME: It looks like hasHashedBookingLink is set to true based on the value of hashedLink when sending the request. So, technically we could remove hasHashedBookingLink usage completely
```
Not fixing this as this is existing behaviour and might have think about it a bit more.

ThyMinimalDev commented on packages/features/bookings/lib/onBookingEvents/BookingEventHandlerService.ts:
```diff
@@ -0,0 +1,70 @@
+  hashedLinkService: HashedLinkService;
+}
+
+export class BookingEventHandlerService {
```
maybe rename to BookingSideEffectService or something like that

ThyMinimalDev commented on packages/features/bookings/lib/service/RegularBookingService.ts:
```diff
@@ -2227,10 +2229,47 @@ async function handler(
+    await bookingEventHandler.onBookingCreated({
+      ...bookingCreatedPayload,
+    });
+  }
```
could be suitable for a Strategy pattern if we plan on adding more if/else cases in the future, more for a followup if we see this keeps growing

hariombalhara commented on packages/features/bookings/lib/onBookingEvents/BookingEventHandlerService.ts:
```diff
@@ -0,0 +1,70 @@
+    }
+  }
+
+  private shouldProcess(payload: BookingCreatedPayload | BookingRescheduledPayload) {
```
Yeah, I was thinking about it, will remove

hariombalhara commented on packages/features/bookings/lib/onBookingEvents/BookingEventHandlerService.ts:
```diff
@@ -0,0 +1,70 @@
+    });
+  }
+
+  private async updatePrivateLinkUsage(payload: BookingCreatedPayload | BookingRescheduledPayload) {
```
Agreed

devin-ai-integration[bot] committed:
refactor: improve ISP compliance in BookingEventHandlerService

- Refactor updatePrivateLinkUsage to only accept hashedLink parameter instead of full payload
- Remove shouldProcess function and inline isDryRun check for better clarity
- Addresses feedback from PR #24025

Co-Authored-By: hariom@cal.com <hariombalhara@gmail.com>

hariombalhara committed:
define only what is used

hariombalhara committed:
Define hashed-link-service as well in api-v2

devin-ai-integration[bot] committed:
fix: export HashedLinkService through platform-libraries

- Add HashedLinkService to platform-libraries/private-links.ts
- Update API V2 to import from platform library instead of direct path
- Fixes MODULE_NOT_FOUND error in API V2 build

Co-Authored-By: hariom@cal.com <hariombalhara@gmail.com>

hariombalhara committed:
Merge branch 'main' into hariom-pri-305-post-booking-handling

alishaz-polymath reviewed (approved):
LGTM 👏

ThyMinimalDev committed:
Merge branch 'main' into hariom-pri-305-post-booking-handling
